proposals/pubsub-mq-interfaces-proposal.md

```diff
@@ -0,0 +1,144 @@
+
+##  1. <a name='TableofContents'></a>Table of Contents
+
+<!-- vscode-markdown-toc -->
+* 1. [Table of Contents](#TableofContents)
+* 2. [Background](#Background)
+	* 2.1. [`message-queue` Background](#message-queueBackground)
+	* 2.2. [`pubsub` Background](#pubsubBackground)
+	* 2.3. [`pubsub`/`message-queue` Comparison](#pubsubmessage-queueComparison)
+* 3. [`pubsub`/`http` Interoptability](#pubsubhttpInteroptability)
+* 4. [Problem Statement](#ProblemStatement)
+* 5. [Proposal](#Proposal)
+	* 5.1. [`pubsub` WIT](#pubsubWIT)
+	* 5.2. [`slight` Runtime](#slightRuntime)
+* 6. [Alternatives](#Alternatives)
+	* 6.1. [Alternative 1](#Alternative1)
+* 7. [Additional Details](#AdditionalDetails)
+
+<!-- vscode-markdown-toc-config
+	numbering=true
+	autoSave=true
+	/vscode-markdown-toc-config -->
+<!-- /vscode-markdown-toc -->
+
+##  2. <a name='Background'></a>Background
+
+###  2.1. <a name='message-queueBackground'></a>`message-queue` Background
+
+`message-queue`s are a messaging mechanism, where message transmitters send messages to a queue construct that can be read by receivers.
+
+![](https://i.imgur.com/Rsa310j.png)
+
+###  2.2. <a name='pubsubBackground'></a>`pubsub` Background
+
+`pubsub` is a messaging pattern, where, instead of sending messages to specific receivers, publishers (i.e., message transmitters) send messages to a channel (or, topic). On the other end, subscribers (i.e., message receivers) can own a subscription to that channel, so they will receive whatever message is sent to it.
+
+![](https://i.imgur.com/pC6SIC0.png)
+
+In addition, there are two ways to receive messages:
+1. Pull-based: The subscriber pulls messages from the channel.
+1. Push-based: The channel pushes messages to the subscriber.
+
+Plus, `pubsub`s usually surface the idea of "Quality of Service" (QoS):
+1. At most once: The message is delivered at most once, but could be lost.
+1. At least once: The message delivery is guaranteed, but could be duplicated.
+1. Exactly once: The message *pretty much* delivered once.
+
+###  2.3. <a name='pubsubmessage-queueComparison'></a>`pubsub`/`message-queue` Comparison
+
+Historically, `pubsub` services (i.e., implementers of the `pubsub` pattern)  are thought to be comprised of non-competitive consumers – That is, if a message is sent to a channel, every receiver that is subscribed to that channel will, eventually, receive it.
+
+While, on the other hand, historically, `message-queue`s are thought to be comprised of competitive consumers – That is, a queued message is dequeued as soon as any one receiver has read it.
+
+However, if we look at services that provide `message-queue` / `pubsub` implementations, the line between the two concepts becomes blurred.
+
+That is, today's most used related services provide both `message-queue`, and `pubsub` capabilities. Below, is a visual example from [HiveMQ](https://www.hivemq.com/) displaying an event-driven architecture with both `message-queue`, and `pubsub` capabilities:
```

Comment:
This section is really talking about the comparison between "topics" and "queues". PubSub is a pattern for asynchronous communication between publishers and subscribers. It does not assume whether the message is sent to one receiver or multiple receivers. Topics and queues, however, are terms used to describe non-competitive vs competitive-consumers. A message in a topic is sent, at the same time, to all subscribes to that topic whereas a message in a queue is sent once to one consumer.

Reply:
Topics = fan-out
queues = load-balancing